### DISCUSSION TITLE
Should we use registerRequestHandler or handle everything in onSessionRequest stream?

### DISCUSSION THREAD

**Discussion by Rimantovas** [WalletConnect/WalletConnectFlutterV2#253] [General]
Currently I have walletconnect implementation done like in the repository example using registerRequestHandler, but when I checked react-wallet-v2 example, it implemented all request handlers in the onSessionRequest stream. Which is the preferred method or is there no difference? 

One thing I found with using registerRequestHandler is that I can't seem to return a valid response back to the dApp. For example, when handling eth_sendTransaction, if the user doesn't approve the request, I can only send back a null transaction hash to produce some sort of error on the dApp's side. If I send anything else, it will handle it as a successful request.

**Comment by quetool** [comment]
Hello @Rimantovas, thanks for reaching out!

> Currently I have walletconnect implementation done like in the repository example using registerRequestHandler, but when I checked react-wallet-v2 example, it implemented all request handlers in the onSessionRequest stream. Which is the preferred method or is there no difference?

Indeed `registerRequestHandler` is the only way on flutter sdk to support methods from dapp. `react-wallet-v2` uses a different approach as it is done with React.JS

> One thing I found with using registerRequestHandler is that I can't seem to return a valid response back to the dApp. For example, when handling eth_sendTransaction, if the user doesn't approve the request, I can only send back a null transaction hash to produce some sort of error on the dApp's side. If I send anything else, it will handle it as a successful request.

The proper way to respond with errors AS A WALLET, is to use `respondSessionRequest()` method from `web3wallet` instance. You can see the Web way of doing it for guidance as the Flutter doc for that part is wrong https://docs.walletconnect.com/web3wallet/wallet-usage?platform=web#responding-to-session-requests

Here there are error codes everyone should follow https://specs.walletconnect.com/2.0/specs/clients/sign/error-codes

The proper way to capture request errors (such as user rejections) AS A DAPP is by listening to relay messages and parsing it content

For instance:
```dart
  _web3App!.core.relayClient.onRelayClientMessage.subscribe(_onRelayMessage);

  void _onRelayMessage(MessageEvent? args) async {
    if (args != null) {
      final payloadString = await _web3App!.core.crypto.decode(
        args.topic,
        args.message,
      );
      final data = jsonDecode(payloadString ?? '{}') as Map<String, dynamic>;
      debugPrint(data.toString());
    }
  }
  ```
  
  It's up to the wallets to send the proper message in the proper format and it's up to the dApps to capture them properly.

**Comment by quetool** [comment]
Hello @Rimantovas! Just looked deepest into Flutter's SDK and you don't have to use `respondSessionRequest()` if you don't want to. as a wallet, if you want to communicate user rejection to the dapp you just need to return a `JsonRpcError`. Here it is a simple pseudo-code for `personal_sign`

```dart
Future<dynamic> personalSignHandler(String topic, dynamic parameters) async {
  // Show approval modal
  if (approved) {
    final signature = // => Sign message code
    return signature;
  } else {
    return const JsonRpcError(code: 5001, message: 'User rejected method');
  }
}
```

**Reply by Rimantovas** [reply]
Thanks, didn't think of that

**Reply by quetool** [reply]
Hello @Rimantovas! I was testing out different web-based dapps and it turns out they behave better if flutter answers as follows:

```dart
Future<dynamic> personalSignHandler(String topic, dynamic parameters) async {
  // Show approval modal
  if (approved) {
    final signature = // => Sign message code
    return _web3Wallet.respondSessionRequest(
      topic: topic,
        response: JsonRpcResponse(
        id: requestId,
        jsonrpc: '2.0',
        result: '0x$signature',
      ),
    );
  } else {
    return _web3Wallet.respondSessionRequest(
      topic: topic,
        response: JsonRpcResponse(
        id: requestId,
        jsonrpc: '2.0',
        error: const JsonRpcError(code: 5001, message: 'User rejected method',)
      ),
    );
  }
}
```

I will include this in Web3Wallet documentation for flutter soon.
To be clear, the previous mentioned way is not wrong, it's just some dapps may not handle it properly
